tlv/varint_test.go

```diff
@@ -0,0 +1,225 @@
+package tlv_test
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"math"
+	"testing"
+
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+type HexBytes []byte
+
+func (b HexBytes) MarshalJSON() ([]byte, error) {
+	return json.Marshal(hex.EncodeToString(b))
+}
+
+type varIntTest struct {
+	Name   string   `json:"name"`
+	Value  uint64   `json:"value"`
+	Bytes  HexBytes `json:"bytes"`
+	ExpErr string   `json:"exp_error,omitempty"`
+}
+
+var writeVarIntTests = []varIntTest{
+	{
+		Name:  "zero",
+		Value: 0x00,
+		Bytes: []byte{0x00},
+	},
+	{
+		Name:  "one byte high",
+		Value: 0xfc,
+		Bytes: []byte{0xfc},
+	},
+	{
+		Name:  "two byte low",
+		Value: 0xfd,
+		Bytes: []byte{0xfd, 0x00, 0xfd},
+	},
+	{
+		Name:  "two byte high",
+		Value: 0xffff,
+		Bytes: []byte{0xfd, 0xff, 0xff},
+	},
+	{
+		Name:  "four byte low",
+		Value: 0x10000,
+		Bytes: []byte{0xfe, 0x00, 0x01, 0x00, 0x00},
+	},
+	{
+		Name:  "four byte high",
+		Value: 0xffffffff,
+		Bytes: []byte{0xfe, 0xff, 0xff, 0xff, 0xff},
+	},
+	{
+		Name:  "eight byte low",
+		Value: 0x100000000,
+		Bytes: []byte{0xff, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00},
+	},
+	{
+		Name:  "eight byte high",
+		Value: math.MaxUint64,
+		Bytes: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+	},
+}
+
+// TestWriteVarInt asserts the behavior of tlv.WriteVarInt under various
+// positive and negative test cases.
+func TestWriteVarInt(t *testing.T) {
+	for _, test := range writeVarIntTests {
+		t.Run(test.Name, func(t *testing.T) {
+			testWriteVarInt(t, test)
+		})
+	}
+}
+
+func testWriteVarInt(t *testing.T, test varIntTest) {
+	var (
+		w   bytes.Buffer
+		buf [8]byte
+	)
+	err := tlv.WriteVarInt(&w, test.Value, &buf)
+	if err != nil {
+		t.Fatalf("unable to encode %d as varint: %v",
+			test.Value, err)
+	}
+
+	if bytes.Compare(w.Bytes(), test.Bytes) != 0 {
+		t.Fatalf("expected bytes: %v, got %v",
+			test.Bytes, w.Bytes())
+	}
+}
+
+var readVarIntTests = []varIntTest{
+	{
+		Name:  "zero",
+		Value: 0x00,
+		Bytes: []byte{0x00},
+	},
+	{
+		Name:  "one byte high",
+		Value: 0xfc,
+		Bytes: []byte{0xfc},
+	},
+	{
+		Name:  "two byte low",
+		Value: 0xfd,
+		Bytes: []byte{0xfd, 0x00, 0xfd},
+	},
+	{
+		Name:  "two byte high",
+		Value: 0xffff,
+		Bytes: []byte{0xfd, 0xff, 0xff},
+	},
+	{
+		Name:  "four byte low",
+		Value: 0x10000,
+		Bytes: []byte{0xfe, 0x00, 0x01, 0x00, 0x00},
+	},
+	{
+		Name:  "four byte high",
+		Value: 0xffffffff,
+		Bytes: []byte{0xfe, 0xff, 0xff, 0xff, 0xff},
+	},
+	{
+		Name:  "eight byte low",
+		Value: 0x100000000,
+		Bytes: []byte{0xff, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00},
+	},
+	{
+		Name:  "eight byte high",
+		Value: math.MaxUint64,
+		Bytes: []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+	},
+	{
+		Name:   "two byte not canonical",
+		Bytes:  []byte{0xfd, 0x00, 0xfc},
+		ExpErr: tlv.ErrVarIntNotCanonical.Error(),
```

Comment:
Can `Error()` be called when the comparison is done?

Reply:
fixed 8d16117d06722c2908f10d37e28d471e398a39a2